sdks/uniswapx-sdk/src/utils/dutchBlockDecay.ts

```diff
@@ -0,0 +1,124 @@
+import { BigNumber } from "ethers";
+
+import { NonlinearDutchDecay } from "../order";
+/*
+These functions mimic the smart contract functions as closely as possible to ensure that the same results are produced.
+Essentially Solidity translated to TypeScript.
+*/
+function locateArrayPosition(
+	curve: NonlinearDutchDecay,
+	targetValue: number
+): [number, number] {
+	const relativeBlocks = curve.relativeBlocks;
+	// eslint-disable-next-line prefer-const
+	let prev = 0;
+	// eslint-disable-next-line prefer-const
+	let next = 0;
+
+	while (next < curve.relativeAmounts.length) {
+		if (relativeBlocks[next] >= targetValue) {
+			return [prev, next];
+		}
+		prev = next;
+		next++;
+	}
+
+	return [next - 1, next - 1];
+}
+
+class NonLinearDutchDecayLib {
+	static decay(
+		curve: NonlinearDutchDecay,
+		startAmount: BigNumber,
+		decayStartBlock: number,
+		currentBlock: number
+	): BigNumber {
+		// mismatch of relativeAmounts and relativeBlocks
+		if (curve.relativeAmounts.length > 16) {
+			throw new Error('InvalidDecayCurve');
+		}
+
+		// handle current block before decay or no decay
+		if (decayStartBlock >= currentBlock || curve.relativeAmounts.length === 0) {
+			return startAmount;
+		}
+
+		const blockDelta = currentBlock - decayStartBlock;
+
+		// Special case for when we need to use the decayStartBlock (0)
+		if (curve.relativeBlocks[0] > blockDelta) {
+			return this.linearDecay(
+				0,
+				curve.relativeBlocks[0],
+				blockDelta,
+				startAmount,
+				startAmount.sub(curve.relativeAmounts[0].toString())
+			);
+		}
+
+		// the current pos is within or after the curve
+		const [prev, next] = locateArrayPosition(curve, blockDelta);
+		//relativeAmounts holds BigInts so we can't directly subtract without conversion
+		const lastAmount = startAmount.sub(curve.relativeAmounts[prev].toString());
+		const nextAmount = startAmount.sub(curve.relativeAmounts[next].toString());
+		return this.linearDecay(
+			curve.relativeBlocks[prev],
+			curve.relativeBlocks[next],
+			blockDelta,
+			lastAmount,
+			nextAmount
+		);
+	}
+
+	static linearDecay(
+		startPoint: number,
+		endPoint: number,
+		currentPoint: number,
+		startAmount: BigNumber,
+		endAmount: BigNumber
+	): BigNumber {
+		if (currentPoint >= endPoint) {
+			return endAmount;
+		}
+
+		const elapsed = BigNumber.from(currentPoint - startPoint);
+		const duration = BigNumber.from(endPoint - startPoint);
+		if (endAmount.lt(startAmount)) {
+			return startAmount.sub(
+				(startAmount.sub(endAmount)).mul(elapsed).div(duration) //muldivdown in contract
+			);
+		} else {
+			return startAmount.add(
+				(endAmount.sub(startAmount)).mul(elapsed).div(duration) //muldivup in contract
+				//TODO: How can we do muldivup in JS?
```

Comment:
muldivup just rounds up the result.. in the latest PR I've gone with `muldivdown` for both:
https://github.com/Uniswap/UniswapX/blob/c0d699883b02668aed1d95077efb4a78ec7d4cfc/src/lib/DutchDecayLib.sol#L94

Reply:
great, this will eliminate an edge case. updated!